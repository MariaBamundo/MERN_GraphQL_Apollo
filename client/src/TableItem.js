import React, {Component} from 'react';
import gql from 'graphql-tag';
import { graphql, Mutation, compose } from 'react-apollo';

class TableItem extends Component
{
    //In the constructor we set up our state and bind our methods to the parent methods in Tasks.js so that
    //when they are called we can change and reflect the data changes in our UI
    constructor(props)
    {
        super(props);
        this.state = {
            item: this.props.task.item,
            id: this.props.task.id,
            isDone: this.props.task.isDone,
            editing: false
        };
    }

    //This method is called when a user hits the save button after editing a task,
    //it updates the items value in the database to reflect the changes
    saveButton()
    {
        this.props.updateItem({
            variables:{
                id: this.state.id,
                item: this.state.item
            },
            update: (store, {data:{updateItem}}) =>
            {
                this.setState({
                    editing: false,
                    item: this.state.item
                });
            }
        });
    }

    //This method is used to update the item state when a user is editing that item
    onChange = (event) =>
    {
        this.setState({item: event.target.value});
    }

    //This method is used to return the correct data that should be displayed in the table, whether it is being edited
    //or not
    createTask()
    {
        if(this.state.editing === true)
        {
            return(
                <div>
                <input value={this.state.item} onChange={this.onChange}/>
                <p></p>
                <button className="save" onClick={()=> this.saveButton()}> Save Item</button>
                </div>
            );
        }
        else
        {
            return(
                <div>
                <p id={this.state.id}>{this.state.item}</p>
                <button className="edit" onClick={()=> this.editButton()}> Edit Item</button>
                </div>
            );
        }
    }

    //This method is called when a user clicks the edit item button and sets the editing state to true
    editButton()
    {
        this.setState({
            editing: true
        });
    }

    render()
    {
        return(this.createTask());
    }
}

const updateItemQuery = gql`
    mutation updateItem($id: String!, $item: String!) {
        updateItem(id: $id, item: $item){
            id
            item
            isDone
        }
    }`;

//At the end we simply export the Item component and the updateItem graphql mutation
export default graphql(updateItemQuery,{name: 'updateItem'})(TableItem);