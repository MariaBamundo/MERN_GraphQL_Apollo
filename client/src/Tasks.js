import React, {Component} from 'react';
import gql from 'graphql-tag';
import {graphql, compose} from 'react-apollo';
import TableItem from './TableItem';

//This class/react component is called from App.js and handles our graphql data changes and the state of the data
//for our item list/table
class Tasks extends Component
{
    //Contructor sets up our props from parent and creates the state of the data for this component
    constructor(props)
    {
        super(props);
        this.state = {
            term: '',
            error: null,
            tasks: this.props.data,
            warning: false
        };
        //Here we manually bind our methods because they do not automatically bind
        this.handleSubmit = this.handleSubmit.bind(this);
        
    }

    //This method is used for when a user enters a new to-do item to our list
    handleSubmit(e) 
    {
        e.preventDefault();
        //First we do some input validation to make sure they have actually typed in an item
        if(this.state.term.replace(/^\s+/g, '').length < 1)
        {
            //If they havent typed in an item to the textbox we display a warning to let them know
            this.setState(
            {
                term: '',
                warning: true
            });
            return;
        }
        else
        {
            //We get here if a non empty string for a to-do list item is entered

            //In this call we are making a mutation call to graphql to add the new to-do item to the database 
            //In the update part of the function we also add the new item to the current Tasks state so it will
            //be reflected in our table
            this.props.addItem(
            {
                variables:
                {
                    item: this.state.term,
                    isDone: false
                },
                update: (store, {data: {addItem}}) => 
                {
                    this.setState({
                        term: '',
                        tasks: [...this.state.tasks,addItem],
                        warning: false
                    });
                }
            }).then(function getResponse(response){
                console.log(response);
            });
        }
    }

    //This method allows for the text inside the text-input box to appear when a user types
    onChange = (event) =>
    {
        this.setState({term: event.target.value});
    }

    //This method is called when a user checks off a task in the UI
    //It updates the isDone value in the database through graphql and updates it in the state aswell so it is reflected
    //in the rendered data
    checkTask(task)
    {
        this.props.updateDone({
            variables:{
                id: task.id,
                item: task.item,
                isDone: !task.isDone
            },
            update: (store, {data:{updateDone}}) =>
            {
                const data = this.state.tasks.slice();
                const index = data.findIndex(task => task.id === updateDone.id);
                data[index] = updateDone;
                this.setState({
                    tasks: data
                });
                //return updateItem;
            }
        });
    }

    //This function is called when a user presses delete on a to-do item, it calls the remove graphql mutation
    //and then removes the item from the state aswell
    deleteTask(task)
    {
        this.props.removeItem(
        {
            variables:
            {
                id: task.id
            },
        }).then(function getResponse(response)
        {
            console.log(response);
        });
        this.setState(prevState => ({
            tasks: prevState.tasks.filter(el => el != task)
        }));
        
    }
    showWarning()
    {
        if(this.state.warning)
        {
            return(<p id="warning"> Please type in a unique item to add it to the To-Do list!</p>)
        }
        else
        {
            return(<p></p>)
        }
    }

    //In this render we create our form for when a user wants to add a new task and 
    //we set up our table and call the TableItem component to fill the table with our components and data
    render()
    {
        return(
            <div className="dataCenter">
                <form className="table" onSubmit={this.handleSubmit}>
                    <input value={this.state.term} onChange={this.onChange}/>
                    {
                        this.showWarning()
                    }
                    <button>Submit</button>
                </form>
                <div className="tableDiv">
                    <table className="table-bordered table-hover">
                        <tbody>
                            {
                                this.state.tasks.map((currentTask) =>
                                <tr key={currentTask.id}>
                                    <td>
                                        <input type="checkbox" checked={currentTask.isDone} className = "checkbox" onChange={() => this.checkTask(currentTask)} />
                                        <TableItem task={currentTask} />
                                        <button className="delete" onClick={()=> this.deleteTask(currentTask)}> Delete Item </button>
                                    </td>
                                </tr>
                                )
                            }
                        </tbody>
                    </table>
                </div>
            </div>
        );
    }
}

//Below are all of the graphql/apollo quieries used above to change the data in the database or return it

const taskQuery = gql`
  query todoList {
    tasks {
      id
      item
      isDone
    }
  }`;

const deleteItem = gql`
    mutation removeItem($id: String!){
        removeItem(id:$id){
            id
            item
            isDone
        }
    }`;

const ADD_TODO = gql`
    mutation addItem($item: String!, $isDone: Boolean!){
      addItem(item: $item, isDone: $isDone){
        id
        item
        isDone
      }
    }`;

const markCompletedQuery = gql`
    mutation updateDone($id: String!, $isDone: Boolean!) {
        updateDone(id: $id, isDone: $isDone){
            id
            item
            isDone
        }
    }`;

//In our export statement we compose a list of the graphql quieries/mutations that we use and also export
//the task component itself
export default compose(graphql(deleteItem,{name: 'removeItem'}),
graphql(taskQuery,{name: 'getTasks'}),
graphql(ADD_TODO,{name: 'addItem'}),graphql(markCompletedQuery,{name: 'updateDone'}))(Tasks);