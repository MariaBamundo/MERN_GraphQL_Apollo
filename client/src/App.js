import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';
import ApolloClient from 'apollo-boost';
import { ApolloProvider, Mutation } from 'react-apollo';
import Tasks from './Tasks';
import gql from 'graphql-tag';

const client = new ApolloClient({
  uri: "http://localhost:4000/graphql",
});

class App extends Component 
{
  constructor(props)
  {
    super(props);
    this.state = {
      term: '',
      error: null,
      data: []
    };
  }


  onSubmit = (event) => 
  {
    event.preventDefault();
    if(this.state.term.replace(/^\s+/g, '').length < 1)
    {
      var warning = document.getElementById('warning');
      warning.style.visibility = "visible";
      this.setState(
      {
        term: ''
      });
            
    }
    else
    {
      //THIS IS WHERE YOU WERE LAST TRYING TO GET THE ADDITEM MUTATION WORKING 
      var warning = document.getElementById('warning');
      warning.style.visibility = "hidden";
      <Mutation mutation={ADD_TODO}>
      {(addItem, {data})=> (
        addItem({variables:{item:this.state.term,isDone:false}})
      )}
      </Mutation>
      this.setState(
      {
        term: ''
      });
    }
  }



  onChange = (event) => 
  {
    this.setState({term: event.target.value});
  }



  render() 
  {
    return(
      <ApolloProvider client={client}>
        <div className="App">
          <h1>To-Do List: </h1>
          <form className="table" onSubmit={this.onSubmit}>
            <input value={this.state.term} onChange={this.onChange}/>
            <p id="warning"> Please type in a unique item to add it to the To-Do list!</p>
            <button>Submit</button>
          </form>
          <div className="tableDiv">
            <table className="table-bordered table-hover">
              <tbody>
                <Tasks />
              </tbody>
            </table>
          </div>
        </div>
      </ApolloProvider>
    );
  }
}

const ADD_TODO = gql`
  mutation addItem($item: String!, $isDone: Boolean!){
    addItem(item: $item, isDone: $isDone){
      id
      item
      isDone
    }
  }
`;

export default App;
