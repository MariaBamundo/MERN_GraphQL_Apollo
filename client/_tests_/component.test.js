import React from 'react';
import App from '../src/App';
import ReactDOM from 'react-dom';
import { shallow, mount, render, configure } from 'enzyme';
import renderer from 'react-test-renderer';
import {ApolloProvider} from 'react-apollo';
import {ApolloClient} from 'apollo-client';
import {HttpLink} from 'apollo-link-http';
import {InMemoryCache} from 'apollo-cache-inmemory';
import fetch from 'node-fetch';
import Tasks from '../src/Tasks';
import TableItem from '../src/TableItem';
import Adapter from 'enzyme-adapter-react-16';
import toJSON from 'enzyme-to-json';




configure({ adapter: new Adapter() });



const client = new ApolloClient({
    link: new HttpLink({
        uri: 'http://localhost:4000/graphql',
        fetch,
    }),
    cache: new InMemoryCache()
});
const tasks = [
      {
        "id": "5b5f145e605ec80c819f5824",
        "item": "one",
        "isDone": false
      }
    ];
describe('Run <App />', () => {
    it('renders our <App /> component', () => {
        const rendered = renderer.create(
            <ApolloProvider client={client}>
                <App />
            </ApolloProvider>
        );
        expect(rendered).toBeTruthy();
    });

    it('Mount our <App /> component', ()=> {
        const appComp = mount(
            <ApolloProvider client={client}>
                <App />
            </ApolloProvider>
        );
        expect(appComp).toBeTruthy();
    });

    it('Testing methods from <Task /> component shallow', () => {
        const taskComponent = shallow(<Tasks data={tasks}/>);
        taskComponent.setState({term:' '});

        let tree = toJSON(taskComponent);
        expect(tree).toMatchSnapshot();
    });
    

});