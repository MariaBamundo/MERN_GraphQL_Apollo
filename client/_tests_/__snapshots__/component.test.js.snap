// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Run <App /> Testing methods from <Task /> component shallow 1`] = `
<Mutation
  ignoreResults={true}
  mutation={
    Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "removeItem",
          },
          "operation": "mutation",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "id",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "removeItem",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "item",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "isDone",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "id",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 134,
        "start": 0,
      },
    }
  }
/>
`;
